---
# Compute the Dependent Variable 'aggressiveLanguage'

The original file for the askscience subreddit has 4 parts- part 1= reading the data, part 2= Compute the 'aggressiveLanguage' variables for every comment, part 3=Compute the 'aggressiveLanguage' variables for each unique user pair, part 4=Update the 'aggressiveLanguage' variables for each comment with the average from each user pair. <br>

This notebook does only part 1 and 2.

**Other than these steps each row needs to be given a timestamp value, which is done after part 1 and part 2**


---

OUTPUT FILE:
1. 'data_result_askscience_subreddits_aggressive_language.csv': Contains the threat, insult, and toxicity scores for each comment<br>


additional details about the data-<br>
the askscience unprocessed data has 26605 comments<br>
there are 13270 comments left after rows with either author as '[deleted]' or body as '[removed]' are removed<br>
there are 7478 user pairs for which there is a network similarity<br>
there are 7478 user pairs for which there is a cultural similarity<br>
thus there should be 7478 unique user pairs<br>
thus for 7952 comments, there is a network similarity and cultural similarity and a parent comment, out of which there are 7478 unique user pairs and 474 repeated user pairs (5.9 percent repeats)


references for detoxify-<br>
https://www.kaggle.com/code/renokan/toxic-comments-using-detoxify-model/notebook?scriptVersionId=87256021 <br>
https://github.com/unitaryai/detoxify <br>



.



.

---
# **Part 1: Reading the data**

In this section, I have read the the output file 'data_fifteen_subreddits_similarity.csv' from the python notebook 'Comment_Level_Network_Cultural_Similarity_askscience.ipynb'. This file contains all the comments for which both network similarity and cultural similarity were computed.

---
.

Check if cuda is being used

In [1]:
import torch
if torch.cuda.is_available():
    device_name = torch.device("cuda")
else:
    device_name = torch.device('cpu')
print("Using {}.".format(device_name))

Using cuda.


Connect to drive

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


reading the file which contains the comments for which both network and cultural similarity, along with parent author was computed

In [3]:
import pandas as pd
import numpy as np
data_similarity_askscience_subreddits = pd.read_csv('/content/gdrive/MyDrive/Colab Notebooks/askscience/data_similarity_askscience_subreddits.csv', low_memory=False, index_col = 0)
print(len(data_similarity_askscience_subreddits)) #length of data = 16370
print(len(pd.unique(data_similarity_askscience_subreddits['subreddit']))) #number of subreddits considered = 1
print(len(pd.unique(data_similarity_askscience_subreddits['id']))) #unique number of comments = , the data is at the comment level =
print(len(pd.unique(data_similarity_askscience_subreddits['parent_id']))) #number of parent nodes =
print(len(pd.unique(data_similarity_askscience_subreddits['link_id']))) #number of submissions =
print(len(pd.unique(data_similarity_askscience_subreddits['author']))) #number of submissions =
print(len(data_similarity_askscience_subreddits.columns))

11782
1
11782
6725
2377
6063
20


In [4]:
data_similarity_askscience_subreddits.head(3)

id   subreddit                                               body  \
5679  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
5681  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
5682  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   

              author  score  gilded  created_utc  retreived_on  permalink  \
5679     omniskeptic      2       0   1664582942    1664960533          0   
5681  jeweledjuniper     11       0   1664583360    1664960508          0   
5682          chop1n      3       0   1664583378    1664960507          0   

      num_comments  url  self_text  is_self parent_id link_id  \
5679             0    0          0        0   iqkee0k  xs73nx   
5681             0    0          0        0   iqke0xc  xs1k1y   
5682             0    0          0        0   iqker6l  xs73nx   

      controversiality  is_submitter  network_similarity  cultural_similarity  \
5679                 0         False            0.985648             0.318494   
5681                 0         False            0.957111             0.642043   
5682                 0         False            0.985648             0.421561   

     parent_comment_author  
5679                chop1n  
5681             feitingen  
5682           omniskeptic

In [5]:
d1 = data_similarity_askscience_subreddits[~data_similarity_askscience_subreddits['parent_comment_author'].isna()]
print(len(d1))

11782


note: this means that 11782 comments have a cultural, network similairty as well as a parent comment

In [6]:
print(data_similarity_askscience_subreddits.columns)

Index(['id', 'subreddit', 'body', 'author', 'score', 'gilded', 'created_utc',
       'retreived_on', 'permalink', 'num_comments', 'url', 'self_text',
       'is_self', 'parent_id', 'link_id', 'controversiality', 'is_submitter',
       'network_similarity', 'cultural_similarity', 'parent_comment_author'],
      dtype='object')


confirm that the following columns do not have missing values

In [7]:
print(data_similarity_askscience_subreddits['network_similarity'].isna().sum())
print(data_similarity_askscience_subreddits['cultural_similarity'].isna().sum())
print(data_similarity_askscience_subreddits['parent_comment_author'].isna().sum())
print(data_similarity_askscience_subreddits['body'].isna().sum())

0
0
0
0


check the unique user-parent value counts

In [8]:
print(len(data_similarity_askscience_subreddits[['author', 'parent_comment_author']].value_counts()))

11100


that means there are 11782-11100 =  682 repeated pairs of user-parent

---
# **Part 2: Compute the 'aggressiveLanguage' variables for every comment**


---
Compute the threatScore and insultScore and toxicityScore for every comment
.

In [9]:
pip install detoxify

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 56.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 79.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 110.8 MB/s eta 0:00:00
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.15.0
    Uninstalling tokenizers-0.15.0:
      Successfully uninstalled tokenizers-0.15.0
  Attempting uninstall: transformers
    Found existing installation: transformers 4.35.2
    Uninstalling transformers-4.35.2:
      Successfully uninstalled transformers-4.35.2


In [10]:
from detoxify import Detoxify

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


Moving 0 files to the new cache system


0it [00:00, ?it/s]

In [11]:
detoxify_model = Detoxify(
    model_type='original',
    device='cuda'
)

Downloading: "https://github.com/unitaryai/detoxify/releases/download/v0.1-alpha/toxic_original-c1212f89.ckpt" to /root/.cache/torch/hub/checkpoints/toxic_original-c1212f89.ckpt
100%|██████████| 418M/418M [00:26<00:00, 16.8MB/s]


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

In [12]:
predicts_dict = detoxify_model.predict("sample sentence")
print(len(predicts_dict))
print(type(predicts_dict))
print(predicts_dict)
print(predicts_dict['threat'])

6
<class 'dict'>
{'toxicity': 0.0006331979, 'severe_toxicity': 0.00012220243, 'obscene': 0.0001922156, 'threat': 0.000112057016, 'insult': 0.00018101634, 'identity_attack': 0.00014015858}
0.000112057016


In [13]:
print(len(data_similarity_askscience_subreddits))
data_similarity_askscience_subreddits.head(3)

11782


id   subreddit                                               body  \
5679  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
5681  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
5682  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   

              author  score  gilded  created_utc  retreived_on  permalink  \
5679     omniskeptic      2       0   1664582942    1664960533          0   
5681  jeweledjuniper     11       0   1664583360    1664960508          0   
5682          chop1n      3       0   1664583378    1664960507          0   

      num_comments  url  self_text  is_self parent_id link_id  \
5679             0    0          0        0   iqkee0k  xs73nx   
5681             0    0          0        0   iqke0xc  xs1k1y   
5682             0    0          0        0   iqker6l  xs73nx   

      controversiality  is_submitter  network_similarity  cultural_similarity  \
5679                 0         False            0.985648             0.318494   
5681                 0         False            0.957111             0.642043   
5682                 0         False            0.985648             0.421561   

     parent_comment_author  
5679                chop1n  
5681             feitingen  
5682           omniskeptic

check that no row has a missing body

In [14]:
data_similarity_askscience_subreddits['body'].isnull().sum()

0

In [15]:
def aggressiveLanguage_mapper(input_data):

  input_data['insult_prob'] = np.nan
  input_data['toxicity_prob'] = np.nan
  input_data['threat_prob'] = np.nan

  ignore_comments_counter = 0
  j = 0

  #additional code to resolve an error
  type_base = type(input_data['parent_id'].iloc[0])

  for ind, row in input_data.iterrows():
    j += 1
    if j % 1000 == 0:
      print('finished comment '+str(j)+'/'+str(len(input_data)))

    curr_author = row['author']
    #curr_subreddit_id = row['subreddit_id'] no subreddit here as there is only one subreddit askscience
    curr_id = row['id']
    curr_comment_body = row['body']

    #check the aggressive language probability in the comment body
    predicts_dict = detoxify_model.predict(curr_comment_body)
    input_data.at[ind,'insult_prob'] = predicts_dict["insult"]
    input_data.at[ind,'toxicity_prob'] = predicts_dict["toxicity"]
    input_data.at[ind,'threat_prob'] = predicts_dict["threat"]

  print('total number of comments ignored: ' +str(ignore_comments_counter))
  return input_data


In [16]:
data_similarity_askscience_subreddits = aggressiveLanguage_mapper(data_similarity_askscience_subreddits)
print(len(data_similarity_askscience_subreddits)) #length of data
print(len(pd.unique(data_similarity_askscience_subreddits['subreddit']))) #number of subreddits considered =
print(len(pd.unique(data_similarity_askscience_subreddits['id']))) #unique number of comments = , the data is at the comment level =
print(len(pd.unique(data_similarity_askscience_subreddits['parent_id']))) #number of parent nodes =
print(len(pd.unique(data_similarity_askscience_subreddits['link_id']))) #number of submissions =
print(len(data_similarity_askscience_subreddits[['author', 'parent_comment_author']].value_counts())) #number of unique counts of speaker-receiver pairs
print(len(data_similarity_askscience_subreddits.columns))

finished comment 1000/11782
finished comment 2000/11782
finished comment 3000/11782
finished comment 4000/11782
finished comment 5000/11782
finished comment 6000/11782
finished comment 7000/11782
finished comment 8000/11782
finished comment 9000/11782
finished comment 10000/11782
finished comment 11000/11782
total number of comments ignored: 0
11782
1
11782
6725
2377
11100
23


In [17]:
print(len(pd.unique(data_similarity_askscience_subreddits['author']))) #number of authors =
data_similarity_askscience_subreddits.head(3)

6063


id   subreddit                                               body  \
5679  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
5681  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
5682  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   

              author  score  gilded  created_utc  retreived_on  permalink  \
5679     omniskeptic      2       0   1664582942    1664960533          0   
5681  jeweledjuniper     11       0   1664583360    1664960508          0   
5682          chop1n      3       0   1664583378    1664960507          0   

      num_comments  ...  parent_id  link_id  controversiality is_submitter  \
5679             0  ...    iqkee0k   xs73nx                 0        False   
5681             0  ...    iqke0xc   xs1k1y                 0        False   
5682             0  ...    iqker6l   xs73nx                 0        False   

     network_similarity  cultural_similarity  parent_comment_author  \
5679           0.985648             0.318494                 chop1n   
5681           0.957111             0.642043              feitingen   
5682           0.985648             0.421561            omniskeptic   

      insult_prob  toxicity_prob threat_prob  
5679     0.000171       0.000725    0.000118  
5681     0.000198       0.000734    0.000142  
5682     0.000175       0.000864    0.000121  

[3 rows x 23 columns]

confirm that there are no missing 'insult', 'threat', and 'toxicity' columns

In [18]:
print(data_similarity_askscience_subreddits['insult_prob'].isnull().sum())
print(data_similarity_askscience_subreddits['threat_prob'].isnull().sum())
print(data_similarity_askscience_subreddits['toxicity_prob'].isnull().sum())

0
0
0


In [19]:
data_similarity_askscience_subreddits.loc[data_similarity_askscience_subreddits.author == '[deleted]', 'author'].count()

0

just a check to ensure that we had removed all these comments in the beginning

In [20]:
data_similarity_askscience_subreddits.head(3)

id   subreddit                                               body  \
5679  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
5681  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
5682  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   

              author  score  gilded  created_utc  retreived_on  permalink  \
5679     omniskeptic      2       0   1664582942    1664960533          0   
5681  jeweledjuniper     11       0   1664583360    1664960508          0   
5682          chop1n      3       0   1664583378    1664960507          0   

      num_comments  ...  parent_id  link_id  controversiality is_submitter  \
5679             0  ...    iqkee0k   xs73nx                 0        False   
5681             0  ...    iqke0xc   xs1k1y                 0        False   
5682             0  ...    iqker6l   xs73nx                 0        False   

     network_similarity  cultural_similarity  parent_comment_author  \
5679           0.985648             0.318494                 chop1n   
5681           0.957111             0.642043              feitingen   
5682           0.985648             0.421561            omniskeptic   

      insult_prob  toxicity_prob threat_prob  
5679     0.000171       0.000725    0.000118  
5681     0.000198       0.000734    0.000142  
5682     0.000175       0.000864    0.000121  

[3 rows x 23 columns]

find number of unique values of the 'created_utc'

In [21]:
print(len(pd.unique(data_similarity_askscience_subreddits['created_utc'])))

11747


In [23]:
print(len(pd.unique(data_similarity_askscience_subreddits['retreived_on'])))

11099


code snippet to convert 'created_utc' timestamp to a readable form

In [24]:
from datetime import datetime
ts = 1664960502
utcDate = datetime.utcfromtimestamp(ts)
print(utcDate.strftime('%Y-%m-%d %H:%M:%S'))
print(utcDate.strftime('%H:%M:%S'))
print(utcDate.strftime('%Y-%m-%d'))
print(utcDate.strftime('%Y-%m-%d:%H'))

2022-10-05 09:01:42
09:01:42
2022-10-05
2022-10-05:09


double check that no column has a missing created timestamp

In [25]:
print(data_similarity_askscience_subreddits['created_utc'].isnull().sum())

0


In [26]:
def assign_comment_timestamp(input_data):

  input_data['date_time'] = np.nan
  input_data['date'] = np.nan
  input_data['time'] = np.nan
  input_data['date_hour'] = np.nan
  input_data['date_hour_min'] = np.nan
  j = 0

  for ind, row in input_data.iterrows():

    j += 1
    if j % 1000 == 0:
      print('finished comment '+str(j)+'/'+str(len(input_data)))

    ts = row['created_utc']
    utcDate = datetime.utcfromtimestamp(ts)
    input_data.at[ind,'date_time'] = utcDate.strftime('%Y-%m-%d_%H:%M:%S')
    input_data.at[ind,'date'] = utcDate.strftime('%Y-%m-%d')
    input_data.at[ind,'time'] = utcDate.strftime('%H:%M:%S')
    input_data.at[ind,'date_hour'] = utcDate.strftime('%Y-%m-%d_%H')
    input_data.at[ind,'date_hour_min'] = utcDate.strftime('%Y-%m-%d_%H:%M')

  return input_data



In [27]:
data_similarity_askscience_subreddits = assign_comment_timestamp(data_similarity_askscience_subreddits)
print(len(data_similarity_askscience_subreddits)) #length of data
print(len(pd.unique(data_similarity_askscience_subreddits['subreddit']))) #number of subreddits considered =
print(len(pd.unique(data_similarity_askscience_subreddits['id']))) #unique number of comments = , the data is at the comment level =
print(len(pd.unique(data_similarity_askscience_subreddits['author']))) #number of author =
print(len(pd.unique(data_similarity_askscience_subreddits['parent_id']))) #number of parent nodes =
print(len(pd.unique(data_similarity_askscience_subreddits['link_id']))) #number of submissions =
print(len(data_similarity_askscience_subreddits[['author', 'parent_comment_author']].value_counts())) #number of unique counts of speaker-receiver pairs
print(len(data_similarity_askscience_subreddits.columns))

finished comment 1000/11782
finished comment 2000/11782
finished comment 3000/11782
finished comment 4000/11782
finished comment 5000/11782
finished comment 6000/11782
finished comment 7000/11782
finished comment 8000/11782
finished comment 9000/11782
finished comment 10000/11782
finished comment 11000/11782
11782
1
11782
6063
6725
2377
11100
28


In [28]:
data_similarity_askscience_subreddits.head(30)

id   subreddit                                               body  \
5679  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
5681  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
5682  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   
5692  iqkpm44  askscience  Thank you for your submission! Unfortunately, ...   
5701  iqkrd5j  askscience  Thats also what I remember. There was speculat...   
5702  iqkre4b  askscience  Not sure if you’re writing only about insects,...   
5706  iqkww4r  askscience  Also your hearing recognizes the tones as cert...   
5709  iqkyg8o  askscience  This is why you feel blinded when youre drivin...   
5713  iqkz2yw  askscience  this sounds very interesting! would love to se...   
5715  iql02a4  askscience  Give me multiverse version of Starship Trooper...   
5721  iql6gj9  askscience  They also have the advantage of being small. L...   
5722  iql6hqk  askscience  Now if only we could harness this claw formati...   
5724  iql3p0f  askscience  PLEASE offer some sources or details on this, ...   
5725  iql4er9  askscience  Thank you for your submission! Unfortunately, ...   
5726  iql4ua2  askscience  Thank you for your submission! Unfortunately, ...   
5728  iql5eg9  askscience  Thank you for your submission! Unfortunately, ...   
5729  iqlb9ha  askscience  Well actually if you dont use glasses for prol...   
5730  iql7kwq  askscience  Im sure youre aware of this, but just in case ...   
5734  iqlc84a  askscience  Not entirely sure, but I think the idea is tha...   
5736  iql95n0  askscience  We always recover our transformations in liqui...   
5737  iql9mk1  askscience  That just reminded me of a line in an Animorph...   
5740  iql9xkx  askscience  Is there a drawback to resilin?It seems pretty...   
5741  iql9ye1  askscience  &gt;There is no universal behavior “programmed...   
5743  iqlajga  askscience  I don’t know, there very well might be some si...   
5750  iqljyko  askscience  The problem with this question is the normal a...   
5751  iqlk4h2  askscience                 What a substantial answer, thanks!   
5754  iqllidm  askscience  No theyre logarithmic. Maybe the most obvious ...   
5755  iqllin1  askscience  Yeah, I knew my ignorance would lead to incons...   
5756  iqlls90  askscience  Oh. So cameras aren’t as good as eyes yet? Bec...   
5759  iqlmfg2  askscience  Thank you for your submission! Unfortunately, ...   

                   author  score  gilded  created_utc  retreived_on  \
5679          omniskeptic      2       0   1664582942    1664960533   
5681       jeweledjuniper     11       0   1664583360    1664960508   
5682               chop1n      3       0   1664583378    1664960507   
5692   askscience-modteam      1       0   1664588425    1664960199   
5701            greese007      2       0   1664589335    1664960145   
5702     viciousfishous08     34       0   1664589349    1664960143   
5706          tin_man6328      4       0   1664592247    1664959974   
5709  yeswehavenotomatoes     34       0   1664593112    1664959927   
5713           mib_sum1ls     13       0   1664593459    1664959907   
5715             glomgore      4       0   1664594017    1664959877   
5721   round-earth-theory     35       0   1664597852    1664959681   
5722       prophetofhelix     12       0   1664597873    1664959680   
5724    nomoremonkeybrain     15       0   1664596152    1664959766   
5725        automoderator      1       0   1664596587    1664959744   
5726        automoderator      1       0   1664596850    1664959730   
5728        automoderator      1       0   1664597195    1664959714   
5729            manu99_it     11       0   1664601059    1664959535   
5730    youareonredditnow      7       0   1664598566    1664959647   
5734                dnick      5       0   1664601747    1664959505   
5736           gertalives      1       0   1664599587    1664959599   
5737      successive-hare      7       

learn more about the time period of the data. We can see thats its a 1 month snapshot. Looks like since 'date' has 31 values it might be a goodvariable for the fixed effect. Also, if we take hour wise, there are 742 different values.

In [29]:
print(len(pd.unique(data_similarity_askscience_subreddits['date_time'])))
print(pd.unique(data_similarity_askscience_subreddits['date_time']))
print(len(pd.unique(data_similarity_askscience_subreddits['date'])))
print(pd.unique(data_similarity_askscience_subreddits['date']))
print(len(pd.unique(data_similarity_askscience_subreddits['time'])))
print(pd.unique(data_similarity_askscience_subreddits['time']))
print(len(pd.unique(data_similarity_askscience_subreddits['date_hour'])))
print(pd.unique(data_similarity_askscience_subreddits['date_hour']))
print(len(pd.unique(data_similarity_askscience_subreddits['date_hour_min'])))
print(pd.unique(data_similarity_askscience_subreddits['date_hour_min']))

11747
['2022-10-01_00:09:02' '2022-10-01_00:16:00' '2022-10-01_00:16:18' ...
 '2022-10-31_19:04:16' '2022-10-31_23:42:20' '2022-10-31_23:42:59']
31
['2022-10-01' '2022-10-30' '2022-10-29' '2022-10-02' '2022-10-03'
 '2022-10-04' '2022-10-05' '2022-10-06' '2022-10-31' '2022-10-07'
 '2022-10-08' '2022-10-12' '2022-10-09' '2022-10-10' '2022-10-11'
 '2022-10-13' '2022-10-14' '2022-10-15' '2022-10-16' '2022-10-17'
 '2022-10-18' '2022-10-19' '2022-10-20' '2022-10-21' '2022-10-22'
 '2022-10-23' '2022-10-24' '2022-10-25' '2022-10-26' '2022-10-27'
 '2022-10-28']
10958
['00:09:02' '00:16:00' '00:16:18' ... '19:04:16' '23:42:20' '23:42:59']
742
['2022-10-01_00' '2022-10-01_01' '2022-10-01_02' '2022-10-01_03'
 '2022-10-01_04' '2022-10-01_05' '2022-10-01_07' '2022-10-01_06'
 '2022-10-01_08' '2022-10-01_09' '2022-10-01_10' '2022-10-01_11'
 '2022-10-01_12' '2022-10-30_15' '2022-10-01_13' '2022-10-01_14'
 '2022-10-01_15' '2022-10-29_18' '2022-10-01_16' '2022-10-01_17'
 '2022-10-01_18' '2022-10-01_19' '

In [30]:
data_similarity_askscience_subreddits.to_csv('/content/gdrive/MyDrive/Colab Notebooks/askscience/data_result_askscience_subreddits_aggressive_language.csv')

ignore from here

In [ ]:
import pandas as pd
data_similarity_askscience_subreddits = pd.read_csv('/content/gdrive/MyDrive/Colab Notebooks/askscience/data_result_askscience_subreddits_aggressive_language.csv',encoding_errors='ignore', index_col=0)

In [ ]:
import pandas as pd
d = pd.read_json('/content/gdrive/MyDrive/Colab Notebooks/askscience/askcience_processed_1.ndjson',encoding_errors='ignore')

In [ ]:
d.head(10)

id   subreddit                                               body  \
0  iqker6l  askscience  No it does not imply that. “We don’t yet know”...   
1  iqkewq0  askscience  while insect muscle might be similar to ours s...   
2  iqkfdmz  askscience                                          [removed]   
3  iqkfl8j  askscience  Pasteurization works by heating (generally a l...   
4  iqkfmj9  askscience  It *absolutely* implies an expectation, even i...   
5  iqkfrm5  askscience  PhD in yeast genetics here, so I’ve streaked t...   
6  iqkfsgy  askscience  Others have given great reasons for why our si...   
7  iqkft4v  askscience                                          [removed]   
8  iqkfzvn  askscience  Inside a living human isn’t lightless dark. Li...   
9  iqkg3r0  askscience  Wordy is good.  your explanation is helping me...   

                author  score  gilded  created_utc parent_id link_id  \
0          omniskeptic      2       0   1664582942   iqkee0k  xs73nx   
1  regular_modern_girl    452       0   1664583016   iqjssf5  xs9pjy   
2            [deleted]      1       0   1664583252   iqkb49u  xs9pjy   
3       jeweledjuniper     11       0   1664583360   iqke0xc  xs1k1y   
4               chop1n      3       0   1664583378   iqker6l  xs73nx   
5        smallwhitedog      4       0   1664583450    xs1k1y  xs1k1y   
6           moewind420      6       0   1664583462    xs73nx  xs73nx   
7            [deleted]      1       0   1664583472    xs1k1y  xs1k1y   
8       sovietamerican      7       0   1664583564   iqk1nsq  xs4rhf   
9            tonytoews      2       0   1664583615   iqk8u6o  xs73nx   

   retrieved_on  controversiality  is_submitter  
0    1664960533                 0         False  
1    1664960528                 0         False  
2    1664960514                 0         False  
3    1664960508                 0         False  
4    1664960507                 0         False  
5    1664960502                 0         False  
6    1664960501                 0         False  
7    1664960501                 0         False  
8    1664960495                 0         False  
9    1664960492                 0         False

In [ ]:
print(d['retrieved_on'].unique())

[1664960533 1664960528 1664960514 ... 1667844629 1667844624 1667844621]


In [ ]:
print(len(d['retrieved_on'].unique()))

23480
